### ISSUE TITLE
Change font in runtime

### ISSUE THREAD

**Issue by Catvert** [Gekkio/imgui-rs#147]
Hello !
I've search to change the font in runtime, with functions like ImGui::PushFont.. But I've found nothing like that.
Is the functionnality implemented ? 
Thanks for your help,
Catvert.

**Comment by malikolivier** [comment]
This function does not have a safe wrapper (yet!). For the time being, you can directly use the unsafe function from the `imgui-sys` crate.

```rust
extern crate imgui_sys;

fn do_something() {
     imgui_sys::igPushFont(my_font_ptr):
     /* .. */
     imgui_sys::igPopFont();
}
```

**Comment by Catvert** [comment]
Hi !
Thanks for your answer. I'll take a look !

**Comment by Catvert** [comment]
Hi !
I have tried to set a font with igPushFont but I can't access to the font field because it's private. I have tried that : 

```rust
let font = ui.imgui().fonts().index_font(0);
unsafe {
  sys::igPushFont(font.font);
}
```

Is there a way to achieve that?

Thanks for your help.

**Comment by malikolivier** [comment]
Thanks for the question.

Sorry I was a bit naive in my previous answer. 
Would the following patch in imgui-rs solve your problem?

```patch
diff --git a/src/fonts.rs b/src/fonts.rs
index 25c63f3..3233406 100644
--- a/src/fonts.rs
+++ b/src/fonts.rs
@@ -239,7 +239,7 @@ impl Default for ImFontConfig {
 
 /// A handle to an imgui font.
 pub struct ImFont<'a> {
-    font: *mut sys::ImFont,
+    pub(crate) font: *mut sys::ImFont,
     _phantom: PhantomData<&'a mut sys::ImFont>,
 }
 impl<'a> ImFont<'a> {
diff --git a/src/lib.rs b/src/lib.rs
index c8461a6..be40535 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1533,6 +1533,14 @@ impl<'ui> Ui<'ui> {
     }
 }
 
+impl <'ui> Ui<'ui> {
+    pub fn with_font<F>(font: &ImFont, f: F) where F: FnOnce() {
+        unsafe { sys::igPushFont(font.font); }
+        f();
+        unsafe { sys::igPopFont(); }
+    }
+}
+
 /// # Utilities
 impl<'ui> Ui<'ui> {
     /// Returns `true` if the last item is being hovered by the mouse.
```

I guess it may not solve your problem yet, as you will probably have mutability issues as is (`fonts()` expects the `ImGui` object to be mutable).
We may need to rework some code to allow to get an immutable reference to a font. Then we will be able to make a safe wrapper around `igPushFont` and `igPopFont`.

**Comment by Catvert** [comment]
Hi !
Thanks for your answer, I've look at your solution and indeed, there is a problem of reference mutability.
I'll wait for a pr.

**Comment by vinterskog** [comment]
Any news here? I mean, it would be nice to be able to change the font and all... ;)

**Comment by wdevore** [comment]
I tried modifying `imgui-rs` as follows but ran into the mutability issues that @malikolivier mentioned:

**font.rs**
```
 ...
 pub fn font(&self) -> *mut sys::ImFont {
        self.font
    }
```

**lib.rs**
```
// Fonts
impl<'ui> Ui<'ui> {
    pub fn push_font(&self, index: usize) {      <== modifying self only cascades the issue. ;-(
        let mut fonts = self.imgui.fonts();      <=== mutability issue. Argh!
        let font = fonts.index_font(index);
        unsafe {
            sys::igPushFont(font.font());
        }
    }

    pub fn pop_font(&self) {
        unsafe {
            sys::igPopFont();
        }
    }
}
```

For now I guess I will use the default font until this issue has been refactored.

Cheers

**Comment by AberrantWolf** [comment]
I worked out a solution over on my fork, but talking with gekkio, it sounds like he's got some font work in a dev branch that (I'm assuming) would make my changes not really work any more, so I'm not sure if it's worth trying to do a whole MR over this until after gekkio's changes come in.

**Comment by Gekkio** [comment]
Initial support for runtime font changes is now implemented in the upcoming 0.1 version. [This example](https://github.com/Gekkio/imgui-rs/blob/b4c31eb65324a4d54ebc592750c48ac50a7b0136/imgui-examples/examples/multiple_fonts.rs) shows how multiple fonts can be used

**Comment by Gekkio** [comment]
Fixed by `0.1.0-rc.1`
